### PR TITLE
add namesuffix for kustomize applications

### PR DESCRIPTION
**Pull request by jeffmhastings**
This PR adds support for setting nameSuffix for kustomize applications.

Addresses https://github.com/argoproj/argo-cd/issues/2301

### CODE REVIEW

File: cmd/argocd/commands/app.go [hunk 1]
```diff
@@ -527,22 +529,23 @@ func setKsonnetOpt(src *argoappv1.ApplicationSource, env *string) {
 	}
 }
 
-func setKustomizeOpt(src *argoappv1.ApplicationSource, namePrefix *string) {
+type kustomizeOpts struct {
```
Comment (alexec) [inline cmd/argocd/commands/app.go]: nice

File: cmd/argocd/commands/app.go [hunk 2]
```diff
@@ -527,22 +529,23 @@ func setKsonnetOpt(src *argoappv1.ApplicationSource, env *string) {
...
+func setKustomizeOpt(src *argoappv1.ApplicationSource, opts kustomizeOpts) {
 	if src.Kustomize == nil {
 		src.Kustomize = &argoappv1.ApplicationSourceKustomize{}
 	}
-	if namePrefix != nil {
-		src.Kustomize.NamePrefix = *namePrefix
-	}
-	if src.Kustomize.IsZero() {
-		src.Kustomize = nil
+	if opts.namePrefix != "" {
```
Comment (alexec) [inline cmd/argocd/commands/app.go]: I don't think you need a guard on this, it is empty string by default

File: cmd/argocd/commands/app.go [hunk 3]
```diff
@@ -527,22 +529,23 @@ func setKsonnetOpt(src *argoappv1.ApplicationSource, env *string) {
...
+		src.Kustomize.NamePrefix = opts.namePrefix
 	}
-}
-func setKustomizeImages(src *argoappv1.ApplicationSource, images []string) {
-	if src.Kustomize == nil {
-		src.Kustomize = &argoappv1.ApplicationSourceKustomize{}
+	if opts.nameSuffix != "" {
```
Comment (alexec) [inline cmd/argocd/commands/app.go]: as above - no need for guard condition

File: test/e2e/fixture/app/actions.go [hunk 4]
```diff
@@ -80,6 +80,11 @@ func (a *Actions) CreateFromFile(handler func(app *Application)) *Actions {
 			NamePrefix: a.context.namePrefix,
 		}
 	}
+	if a.context.nameSuffix != "" {
```
Comment (alexec) [inline test/e2e/fixture/app/actions.go]: what happens if you have both suffix and prefix - doesn't this replace the prefix?

File: test/e2e/kustomize_test.go [hunk 5]
```diff
@@ -180,3 +180,16 @@ func TestKustomizeImages(t *testing.T) {
 			assert.Contains(t, app.Spec.Source.Kustomize.Images, KustomizeImage("alpine:bar"))
 		})
 }
+
+// make sure we we can invoke the CLI to set namesuffix
+func TestKustomizeNameSuffix(t *testing.T) {
+	Given(t).
```
Comment (alexec) [inline test/e2e/kustomize_test.go]: good stuff

**Review by alexec — COMMENTED** [review]
Good stuff! Well done for updating the docs and the e2e test. Few first-time contributors managed to tick all the boxes! My big query - what happens with both prefix and suffix?

File: test/e2e/fixture/app/actions.go (hunk 4 above)
Comment (jeffmhastings) [reply test/e2e/fixture/app/actions.go]: yeah that's a good point, I'll rework this bit and update the tests

Comment (alexec) [reply test/e2e/fixture/app/actions.go]: I felt like wasn’t the only place. Need to check the other code.

Comment (jeffmhastings) [reply test/e2e/fixture/app/actions.go]: I've double checked, and I don't see any potential issues for cases where nameprefix and namesuffix are both set. Did you see anything else in particular that you were concerned about?

**Review by alexec — APPROVED** [review]
LGTM. Will merge when we've branched v1.3

**Pull request merged**